cpp/include/raft/core/mdspan.hpp

```diff
@@ -22,4 +15,212 @@
+ *          raft::host_mdspan/raft::device_mdspan or their derived types
+ */
+template <typename... Tn>
+inline constexpr bool is_mdspan_v = std::conjunction_v<is_mdspan_t<Tn>...>;
+
+template <typename... Tn>
+using enable_if_mdspan = std::enable_if_t<is_mdspan_v<Tn...>>;
+
+// uint division optimization inspired by the CIndexer in cupy.  Division operation is
+// slow on both CPU and GPU, especially 64 bit integer.  So here we first try to avoid 64
+// bit when the index is smaller, then try to avoid division when it's exp of 2.
+template <typename I, typename IndexType, size_t... Extents>
+MDSPAN_INLINE_FUNCTION auto unravel_index_impl(
+  I idx, std::experimental::extents<IndexType, Extents...> shape)
+{
+  constexpr auto kRank = static_cast<int32_t>(shape.rank());
+  std::size_t index[shape.rank()]{0};  // NOLINT
+  static_assert(std::is_signed<decltype(kRank)>::value,
+                "Don't change the type without changing the for loop.");
+  for (int32_t dim = kRank; --dim > 0;) {
+    auto s = static_cast<std::remove_const_t<std::remove_reference_t<I>>>(shape.extent(dim));
+    if (s & (s - 1)) {
+      auto t     = idx / s;
+      index[dim] = idx - t * s;
+      idx        = t;
+    } else {  // exp of 2
+      index[dim] = idx & (s - 1);
+      idx >>= detail::popc(s - 1);
+    }
+  }
+  index[0] = idx;
+  return detail::arr_to_tup(index);
+}
+
+/**
+ * @brief Create a raft::mdspan
+ * @tparam ElementType the data type of the matrix elements
+ * @tparam IndexType the index type of the extents
+ * @tparam LayoutPolicy policy for strides and layout ordering
+ * @tparam is_host_accessible whether the data is accessible on host
+ * @tparam is_device_accessible whether the data is accessible on device
+ * @param ptr Pointer to the data
+ * @param exts dimensionality of the array (series of integers)
+ * @return raft::mdspan
+ */
+template <typename ElementType,
+          typename IndexType        = std::uint32_t,
+          typename LayoutPolicy     = layout_c_contiguous,
+          bool is_host_accessible   = false,
+          bool is_device_accessible = true,
+          size_t... Extents>
+auto make_mdspan(ElementType* ptr, extents<IndexType, Extents...> exts)
+{
+  using accessor_type = detail::accessor_mixin<std::experimental::default_accessor<ElementType>,
+                                               is_host_accessible,
+                                               is_device_accessible>;
+
+  return mdspan<ElementType, decltype(exts), LayoutPolicy, accessor_type>{ptr, exts};
+}
+
+/**
+ * @brief Create raft::extents to specify dimensionality
+ *
+ * @tparam IndexType The type of each dimension of the extents
+ * @tparam Extents Dimensions (a series of integers)
+ * @param exts The desired dimensions
+ * @return raft::extents
+ */
+template <typename IndexType, typename... Extents, typename = ensure_integral_extents<Extents...>>
+auto make_extents(Extents... exts)
```

Comment:
It looks like this does the same thing as `dextents`.  If so, would you consider just using that?

Reply:
Created for this: https://github.com/rapidsai/raft/issues/818